staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -1369,23 +1370,37 @@ func hasServedCRDVersion(spec *apiextensionsv1.CustomResourceDefinitionSpec, ver
 // buildOpenAPIModelsForApply constructs openapi models from any validation schemas specified in the custom resource,
 // and merges it with the models defined in the static OpenAPI spec.
 // Returns nil models ifthe static spec is nil, or an error is encountered.
-func buildOpenAPIModelsForApply(staticOpenAPISpec *spec.Swagger, crd *apiextensionsv1.CustomResourceDefinition) (*spec.Swagger, error) {
+func buildOpenAPIModelsForApply(staticOpenAPISpec map[string]*spec.Schema, crd *apiextensionsv1.CustomResourceDefinition) (*spec3.OpenAPI, error) {
 	if staticOpenAPISpec == nil {
 		return nil, nil
 	}
 
-	specs := []*spec.Swagger{}
+	// Convert static spec to V3 format to be able to merge
+	staticSpecV3 := &spec3.OpenAPI{
+		Version: "3.0.0",
+		Info: &spec.Info{
+			InfoProps: spec.InfoProps{
+				Title:   "Kubernetes CRD Swagger",
+				Version: "v0.1.0",
+			},
+		},
+		Components: &spec3.Components{
+			Schemas: staticOpenAPISpec,
+		},
+	}
+
+	specs := []*spec3.OpenAPI{staticSpecV3}
 	for _, v := range crd.Spec.Versions {
 		// Defaults are not pruned here, but before being served.
 		// See flag description in builder.go for flag usage
-		s, err := builder.BuildOpenAPIV2(crd, v.Name, builder.Options{V2: true, SkipFilterSchemaForKubectlOpenAPIV2Validation: true, StripValueValidation: true, StripNullable: true, AllowNonStructural: false})
+		s, err := builder.BuildOpenAPIV3(crd, v.Name, builder.Options{})
 		if err != nil {
 			return nil, err
 		}
 		specs = append(specs, s)
 	}
 
-	mergedOpenAPI, err := builder.MergeSpecs(staticOpenAPISpec, specs...)
+	mergedOpenAPI, err := builder.MergeSpecsV3(specs...)
```

Comment:
out of curiosity, is this making a copy of anything?

Reply:
no just makes a new map and combines keys

Reply:
The MergeSpecsV3 function was designed to only operate on CRDs and does not handle conflicts because it only operates on one GV. I did notice that the v2 MergeSpecs function has additional handling for conflicts and priority of the static spec that is not considered in v3 (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/merge.go#L59-L63). Just want to double check if that missing functionality is needed for V3 for this use case.

Reply:
@Jefftree mergespecsV3 has conflict detection and error throwing in its codepath. Is this what you are referring to?

https://github.com/kubernetes/kubernetes/blob/2eb2c88b1d32b7e278787a2c2661072923fcb93b/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/merge.go#L126-L131

I'm curious how definitions other than the builtin meta/apiextensions types and the CRDs from here might be expected to appear in the definitions list. I would expect the only additional definitions to be added to be the CRD and its ListSchemas, both of which incorporate the GVK in the name (so no conflicts across versions):

https://github.com/kubernetes/kubernetes/blob/48e968d2e078049ef125cb437a319208c4dd5de4/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/builder.go#L538-L544

To incorporate for the static spec's priority, I made sure to include it first in the list of `specs` to be merged, which merges in the order provided, and throws errors for duplicate keys.